core/sort.cpp

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#include <numeric>
+
+#include "scipp/core/except.h"
+#include "scipp/core/indexed_slice_view.h"
+#include "scipp/core/sort.h"
+#include "scipp/core/tag_util.h"
+
+namespace scipp::core {
+
+template <class T> struct MakePermutation {
+  static auto apply(const VariableConstProxy &key) {
+    if (key.dims().ndim() != 1)
+      throw except::DimensionError("Sort key must be 1-dimensional");
+
+    // Variances are ignored for sorting.
+    const auto &values = key.values<T>();
+
+    std::vector<scipp::index> permutation(values.size());
+    std::iota(permutation.begin(), permutation.end(), 0);
+    std::sort(
+        permutation.begin(), permutation.end(),
+        [&](scipp::index i, scipp::index j) { return values[i] < values[j]; });
+    return permutation;
+  }
+};
+
+auto makePermutation(const VariableConstProxy &key) {
+  return CallDType<double, float, int64_t, int32_t, bool,
+                   std::string>::apply<MakePermutation>(key.dtype(), key);
+}
+
+Variable sort(const VariableConstProxy &var, const VariableConstProxy &key) {
+  return concatenate(
+      IndexedSliceView{var, key.dims().inner(), makePermutation(key)});
+}
+
+DataArray sort(const DataConstProxy &array, const VariableConstProxy &key) {
+  return concatenate(
+      IndexedSliceView{array, key.dims().inner(), makePermutation(key)});
+}
+
+DataArray sort(const DataConstProxy &array, const Dim &key) {
+  return sort(array, array.coords()[key]);
+}
+
+DataArray sort(const DataConstProxy &array, const std::string &key) {
+  return sort(array, array.labels()[key]);
+}
+
+Dataset sort(const DatasetConstProxy &dataset, const VariableConstProxy &key) {
+  return concatenate(
+      IndexedSliceView{dataset, key.dims().inner(), makePermutation(key)});
+}
```

Comment:
Code triplication with lines 36 - 45. Maybe template this and specialize for other cases, not sure.

Reply:
I think templating is not so easy, without triplicating the amount of *generated* code (compiler would generate code for `Variable`, `VariableProxy`, `VariableConstProxy` (etc.), instead of just `VariableConstProxy`. Of course we could use `std::enable_if`, but I feel this is making the code harder to understand than copying a single line of code. Do you have a concrete suggestion though?